### PR TITLE
fix: optimize Range parsing and formatting

### PR DESCRIPTION
**Pull request by jviide**
This pull request optimizes the Range class in the following ways:

 1. Produce fewer intermediate objects when reducing a range's space characters to single spaces. This seems to improve bench-subset scores by up to 5%, and bench-satisfies scores to a lesser degree.
 2. Optimize Range formatting with explicit for loops instead, avoiding an intermediate array creation. This seems to improve bench-satisfies and bench-subset scores by up to 20%.
 3. Calculate Range's `.range` string (used by `.format()` and `.toString()`) lazily. This seems to improve bench-satisfies and bench-subset scores by up to 9%.

The external interface for the class stays the same, except for the new internal `.formatted` property used to cache its lazily calculated string. `Range#range` property is now also read-only.

There is a new test lazy formatting to ensure full test coverage.

The benchmarks bench-satisfies and bench-subset benefit from these changes, sometimes by up to 40%. Other benchmark results seem to stay the same. Here are the affected benchmarks before:

```
$ node benchmarks/bench-satisfies.js
satisfies(1.0.6, 1.0.3||^2.0.0) x 695,094 ops/sec ±0.68% (97 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0) x 764,115 ops/sec ±0.40% (99 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0) x 805,593 ops/sec ±0.62% (97 runs sampled)
satisfies(1.0.6, 1.0.3||^2.0.0, {"includePrelease":true}) x 695,045 ops/sec ±0.73% (95 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0, {"includePrelease":true}) x 750,433 ops/sec ±0.66% (99 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0, {"includePrelease":true}) x 787,903 ops/sec ±0.39% (99 runs sampled)
satisfies(1.0.6, 1.0.3||^2.0.0, {"includePrelease":true,"loose":true}) x 652,166 ops/sec ±0.34% (99 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0, {"includePrelease":true,"loose":true}) x 696,377 ops/sec ±0.36% (96 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0, {"includePrelease":true,"loose":true}) x 721,729 ops/sec ±0.35% (98 runs sampled)
satisfies(1.0.6, 1.0.3||^2.0.0, {"includePrelease":true,"loose":true,"rtl":true}) x 585,692 ops/sec ±0.75% (95 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0, {"includePrelease":true,"loose":true,"rtl":true}) x 631,653 ops/sec ±0.33% (96 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0, {"includePrelease":true,"loose":true,"rtl":true}) x 650,110 ops/sec ±0.64% (95 runs sampled)

$ node benchmarks/bench-subset.js
subset(1.2.3, *) x 633,342 ops/sec ±0.61% (95 runs sampled)
subset(^1.2.3, *) x 743,036 ops/sec ±0.47% (97 runs sampled)
subset(^1.2.3-pre.0, *) x 680,087 ops/sec ±0.76% (98 runs sampled)
subset(^1.2.3-pre.0, *) x 680,948 ops/sec ±0.46% (96 runs sampled)
subset(1 || 2 || 3, *) x 330,669 ops/sec ±0.53% (98 runs sampled)
```

And after:

```
$ node benchmarks/bench-satisfies.js
satisfies(1.0.6, 1.0.3||^2.0.0) x 896,936 ops/sec ±0.53% (94 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0) x 998,214 ops/sec ±0.40% (95 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0) x 1,000,593 ops/sec ±0.43% (97 runs sampled)
satisfies(1.0.6, 1.0.3||^2.0.0, {"includePrelease":true}) x 890,369 ops/sec ±0.41% (100 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0, {"includePrelease":true}) x 977,239 ops/sec ±0.48% (97 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0, {"includePrelease":true}) x 983,682 ops/sec ±0.95% (96 runs sampled)
satisfies(1.0.6, 1.0.3||^2.0.0, {"includePrelease":true,"loose":true}) x 805,330 ops/sec ±0.84% (98 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0, {"includePrelease":true,"loose":true}) x 894,117 ops/sec ±0.43% (99 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0, {"includePrelease":true,"loose":true}) x 911,742 ops/sec ±0.42% (96 runs sampled)
satisfies(1.0.6, 1.0.3||^2.0.0, {"includePrelease":true,"loose":true,"rtl":true}) x 741,254 ops/sec ±0.35% (97 runs sampled)
satisfies(1.0.6, 2.2.2||~3.0.0, {"includePrelease":true,"loose":true,"rtl":true}) x 807,380 ops/sec ±0.42% (99 runs sampled)
satisfies(1.0.6, 2.3.0||<4.0.0, {"includePrelease":true,"loose":true,"rtl":true}) x 820,390 ops/sec ±0.37% (99 runs sampled)

$ node benchmarks/bench-subset.js
subset(1.2.3, *) x 905,030 ops/sec ±0.63% (96 runs sampled)
subset(^1.2.3, *) x 1,026,457 ops/sec ±0.63% (95 runs sampled)
subset(^1.2.3-pre.0, *) x 923,789 ops/sec ±0.41% (97 runs sampled)
subset(^1.2.3-pre.0, *) x 923,136 ops/sec ±0.44% (96 runs sampled)
subset(1 || 2 || 3, *) x 432,037 ops/sec ±0.67% (96 runs sampled)
```

### CODE REVIEW

File: classes/range.js [hunk 1]
```diff
@@ -29,10 +29,11 @@ class Range {
     // First reduce all whitespace as much as possible so we do not have to rely
     // on potentially slow regexes like \s*. This is then stored and used for
     // future error messages as well.
```
Comment (kurtextrem) [inline classes/range.js:31]: probably worth looking up why the comment has made it into the source (was there a bench or sth?), and if the reason no longer makes sense today, probably worth removing.

File: classes/range.js [hunk 2]
```diff
@@ -29,10 +29,11 @@ class Range {
...
-    this.raw = range
-      .trim()
-      .split(/\s+/)
-      .join(' ')
+    this.raw = range.replace(/\s+/g, ' ')
```
Comment (kurtextrem) [inline classes/range.js:32]: does it make sense to hoist the regex?

File: classes/range.js [hunk 3]
```diff
@@ -66,14 +67,19 @@ class Range {
...
 
-    this.format()
+    this.formatted = undefined
+  }
+
+  get range () {
+    if (this.formatted === undefined) {
+      this.formatted = this.set
+        .map((comps) => comps.join(' ').trim())
+        .join('||')
```
Comment (kurtextrem) [inline classes/range.js:77]: How often is `range` usually called, would it make sense to replace this with a for loop too?

File: classes/range.js (hunk 1 above)
Comment (jviide) [reply classes/range.js:31]: The code's functionality has not really changed since the comment was added in #564 by @lukekarrys. As such I'm not able to comment. However, I think investigating it falls somewhat outside the scope of this pull request.

File: classes/range.js (hunk 2 above)
Comment (jviide) [reply classes/range.js:32]: I did benchmark a the hoisted version prior to submitting this pull request, and it did not have a measurable impact in my environment. Hoisting won't hurt here, though, I'll do that 👍

File: classes/range.js (hunk 1 above)
Comment (kurtextrem) [reply classes/range.js:33]: My intention was rather to figure out if we should remove it, because it says "...so we do not have to rely on potentially slow regexes like \s*". In this PR we switch to a regex, so the comment feels slightly out of place.

File: classes/range.js (hunk 3 above)
Comment (jviide) [reply classes/range.js:75]: That's a good question! The benchmarks don't seem to contain anything that get the `.range` property. Transforming this into loops like this seems to speed up the calculation considerably though:

```js
  get range () {
    if (this.formatted === undefined) {
      let formatted = ''
      for (let i = 0; i < this.set.length; i++) {
        if (i > 0) {
          formatted += '||'
        }
        const comps = this.set[i]
        for (let k = 0; k < comps.length; k++) {
          if (k > 0) {
            formatted += ' '
          }
          formatted += comps[k].toString().trim()
        }
      }
      this.formatted = formatted
    }
    return this.formatted
  }
```

File: classes/range.js (hunk 1 above)
Comment (jviide) [reply classes/range.js:33]: The previous version also used a similar regex:

```js
this.raw = range
  .trim()
  .split(/\s+/)
  .join(' ')
```

It was introduced in the same pull request as the comment: https://github.com/npm/node-semver/pull/564/files#diff-95bd101ea814ee8011aedabd5c1d1b1a468dbfbf774168e3c3e319a332deadb6R29-R35

Comment (jviide) [reply classes/range.js:33]: Ah, I think I got it! The same pull request introduces this change to the regular expressions used elsewhere in the library: https://github.com/npm/node-semver/pull/564/files#diff-7e884a3dd192e2887678190dd2acabc002cd4ecb2e7b5801d33ffa5248b98302R12-R28

That's why the whitespaces need to get collapsed, and why `\s*` regexps can be later avoided.

Comment (jviide) [reply classes/range.js:33]: As such I think the comment still holds, as it's not referring to the `\s+` regexp used immediate below it (?)

File: classes/range.js (hunk 3 above)
Comment (jviide) [reply classes/range.js]: I added this in a new commit that's applied before the lazy calculation commit. By itself this change seemed to boost bench-satisfies and bench-subset scores by up to 20%. Thanks for the suggestion! 👍

Comment (kurtextrem) [reply classes/range.js]: Nice win!

**Comment by jviide** [comment]
The "refactor:" pull request name prefix was not allowed so I picked "fix:", and changed the commit messages to use the same prefix as well.

**Comment by wraithgar** [comment]
We generally squash-merge external PRs unless there's a compelling reason to keep each commit.  linting will pass if *either* the PR title or every commit has a valid conventional prefix.

Don't worry about keeping the commit messages "valid" as we go through this PR, as we'll be using the PR title if and when this lands.

File: classes/range.js [hunk 4]
```diff
@@ -1,3 +1,5 @@
+const SPACE_CHARACTERS = /\s+/g
```
Comment (wraithgar) [inline classes/range.js:1]: This is only used once, there's no need to abstract it to a variable.

File: functions/inc.js [hunk 5]
```diff
@@ -9,7 +9,7 @@ const inc = (version, release, options, identifier, identifierBase) => {
 
   try {
     return new SemVer(
-      version instanceof SemVer ? version.version : version,
+      version instanceof SemVer ? version.raw : version,
```
Comment (wraithgar) [inline functions/inc.js:12]: This change seems unrelated to the rest of the PR and may represent a subtly breaking change.  At the very least it needs to be a separate PR where its merits can be explored.

File: test/classes/range.js [hunk 6]
```diff
@@ -82,6 +82,13 @@ test('tostrings', (t) => {
   t.end()
 })
 
+test('format', (t) => {
```
Comment (wraithgar) [inline test/classes/range.js:85]: Future devs will probably appreciate a more descriptive test here.  We know today it's for coverage of the `this.formatted` test but that's not apparent from just this test alone.

**Comment by wraithgar** [comment]
Thanks for the help in reviewing this @kurtextrem.  Yes, that comment was referring to `s*` which is a classic ReDOS vector.  Historically we've used the `.split` approach to fix that, because it's rarely in a hot path where performance matters as much as it does here.   Good to know that `String.replace` is faster.

File: classes/range.js (hunk 4 above)
Comment (marvinhagemeister) [reply classes/range.js:1]: FYI: This is a performance optimization, not a stylistic choice. Reconstructing a new `Regex` instance each time the `Range` class is instantiated is much more expensive then reusing the same one.

File: classes/semver.js [hunk 7]
```diff
@@ -73,14 +73,20 @@ class SemVer {
     }
 
     this.build = m[5] ? m[5].split('.') : []
-    this.format()
+    this.formatted = undefined
```
[the remark this answers is not in the document]
Comment (marvinhagemeister) [reply classes/semver.js:76]: This is not a stylistic choice, but a performance optimization. By declaring all properties upfront, even if they are initialized to `undefined` you can avoid a shape transition in the JIT. That's what's done here.would make the code

File: test/classes/range.js (hunk 6 above)
Comment (marvinhagemeister) [reply test/classes/range.js:85]: To me this seems fine. The test is all about `.format()` calls.

File: functions/inc.js (hunk 5 above)
Comment (jviide) [reply functions/inc.js]: This change was originally described as follows:

> a modification to inc to use the SemVer#raw property instead of SemVer#version to avoid extra formatting work when possible

As such this was a part of the original optimizations for the SemVer class. When a new SemVer object was created based on an older SemVer object, it forced the older SemVer instance to calculate its the `.version` property. To avoid this using the older SemVer instance's `.raw` property (which is the original string used to initialize the older SemVer instance) seemed reasonable.

I reverted this change, which caused the test "increment versions test" to start failing, specifically the "didn't modify its input" part. This happens because `version instanceof SemVer ? version.version : version` line can force the new `.formatted` property to change, technically modifying the SemVer instance used as the input.

Time constraints don't allow me to dive into the tests with confidence, so I removed the SemVer class related optimizations from this pull request.

File: test/classes/range.js (hunk 6 above)
Comment (jviide) [reply test/classes/range.js]: Renamed the test to "formatted value is calculated lazily and cached" and modified it to examine the `.formatted` property more directly.

File: classes/semver.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jviide) [reply classes/semver.js]: The same reasoning applies for the similar changes in range.js (as this PR doesn't modify semver.js anymore).

**Review by wraithgar — APPROVED** [review]
